app/models/client/models/ChatMessage.js

```diff
@@ -9,3 +27,101 @@ ChatMessage.setReactions = function(messageId, reactions) {
 ChatMessage.unsetReactions = function(messageId) {
 	return this.update({ _id: messageId }, { $unset: { reactions: 1 } });
 };
+
+const normalizeThreadMessage = (message) => {
+	if (message.msg) {
+		return renderMessageBody(message).replace(/<br\s?\\?>/g, ' ');
+	}
+
+	if (message.attachments) {
+		const attachment = message.attachments.find((attachment) => attachment.title || attachment.description);
+
+		if (attachment && attachment.description) {
+			return s.escapeHTML(attachment.description);
+		}
+
+		if (attachment && attachment.title) {
+			return s.escapeHTML(attachment.title);
+		}
+	}
+};
+
+const upsertMessage = async ({ msg, subscription, uid = Tracker.nonreactive(() => Meteor.userId()) }, collection = ChatMessage) => {
+	const userId = msg.u && msg.u._id;
+
+	if (subscription && subscription.ignored && subscription.ignored.indexOf(userId) > -1) {
+		msg.ignored = true;
+	}
+
+	if (msg.t === 'e2e' && !msg.file) {
+		msg.e2e = 'pending';
+	}
+	msg = await promises.run('onClientMessageReceived', msg) || msg;
+
+	const { _id, ...messageToUpsert } = msg;
+
+	if (msg.tcount) {
+		collection.direct.update({ tmid: _id }, {
+			$set: {
+				following: msg.replies && msg.replies.indexOf(uid) > -1,
+				threadMsg: normalizeThreadMessage(messageToUpsert),
+				repliesCount: msg.tcount,
+			},
+		}, { multi: true });
+	}
+
+	return collection.direct.upsert({ _id }, messageToUpsert);
+};
+
+function upsertMessageBulk({ msgs, subscription }, collection = ChatMessage) {
+	const uid = Tracker.nonreactive(() => Meteor.userId());
+	const { queries } = ChatMessage;
+	collection.queries = [];
+	msgs.forEach((msg, index) => {
+		if (index === msgs.length - 1) {
+			ChatMessage.queries = queries;
+		}
+		upsertMessage({ msg, subscription, uid }, collection);
+	});
+}
+
+const messagePreFetch = () => {
+	let messagesFetched = false;
+	Tracker.autorun(() => {
+		if (!messagesFetched && CachedChatSubscription.ready.get()) {
+			const status = Meteor.status();
+			if (status.status !== 'connected') {
+				return;
+			}
+			messagesFetched = true;
+			const subscriptions = ChatSubscription.find(
+				{
+					open: true,
+				},
+				{
+					fields: {
+						rid: 1,
+						ls: 1,
+					},
+				},
+			);
+			const limit = parseInt(getConfig('roomListLimit')) || 50;
+			subscriptions.forEach((subscription) => {
+				const ts = undefined;
+				const { rid, ls } = subscription;
+				Meteor.call('loadHistory', rid, ts, limit, ls, (err, result) => {
```

Comment:
This calls can reach the rate limiter really easy, needs to deal with this case

Reply:
What about modifying it to fetch data for all rooms with one call?

Reply:
It may be better, but we don't have an API for that.

Reply:
Load History was made for prefetch, I will refactor it.

Reply:
@rodrigok I see fetching all data for rooms increase the response time and degrades the PWA experience. So what if we fetch for first 10-15 rooms by activity, configurable as an admin setting  for all users?

Reply:
@bharathkkb If we have that configurable as a setting it would be the best option here.

Reply:
@rodrigok I assume wrong handle? 😅

Reply:
@rodrigok number of room data to be fetched is an admin setting now

Reply:
@bharathkkb sorry about that :)

@bhardwajaditya that's great